### ISSUE TITLE
In HEAD, the "math" built-in doesn't handle equality tests. 

### ISSUE THREAD

**Issue by mojotx** [fish-shell/fish-shell#5374]
With fish v.2.7.1:

```fish
fish --version ; math '1 < 2' ; and echo true; or echo false                                                                                                                                                                                    Thu Nov 29 17:18:00 2018
fish, version 2.7.1
1
true
```
With the latest Git HEAD:
```fish
fish --version; math '1 < 2' ; and echo true; or echo false
fish, version 2.7.1-1761-gf0aa63cc
math: Error: Missing operator
'1 < 2'
   ^
false
```

It looks like several months ago @krader1961 [removed the `math.fish` function version](https://github.com/fish-shell/fish-shell/commit/d10decabda8aaeec97c07096a6f7f6f220dc0950), leaving only the built-in, and the built-in uses a slightly tweaked version of https://github.com/codeplea/tinyexpr, and TinyExpr doesn't implement all of the functionality of `bc`, such as comparison operations. 

I'm new to fish shell, so I thought I would post it here for discussion before investing any time in trying to address the issue. The obvious solution would be to extend TinyExpr to handle comparisons, but you all may disagree.  🙂

**Comment by floam** [comment]
I think we should want `math` to be able to do number comparisons.

**Comment by mojoTX** [comment]
I only noticed this issue because of an oh-my-fish plugin that uses math for numeric comparisons, and it started giving me errors when I started running the latest HEAD. In my specific situation, it would be easy to modify the script to use `test` to do the comparison, since it's just comparing integers. I just checked, and using `test` will work for both the current 2.7.1 release and the current git HEAD. 

However, our old `math.fish` function was a wrapper around `bc`, and `bc` has the ability to compare floating point numbers, which is something that the test built-in does not currently do. 

So....do we extend the math built-in for parity with `bc`? Do we extend the test built-in to implement floating-point comparisons? 

If we eliminate the comparisons from math, do we issue a helpful error message suggesting they use test, or use bc directly? Or do we bring back the `math.fish` function?

**Comment by mojoTX** [comment]
I just read all of the discussion for https://github.com/fish-shell/fish-shell/issues/3157, so it appears that this particular horse has already been beat to death.   😂

With bash, you can do integer comparisons with `test`, and if you want to get fancy and do floating-point comparisons, you use `bc` (or something similar). 

So, I don't think this is a big deal, although we should probably document the difference, and perhaps issue a more helpful error message than "missing operator." 

Another idea might be to have the math built-in detect the integer comparison, and just call the test built-in. We could issue a warning that integer comparisons in math are deprecated, and remove them completely in 4.0. 

Thoughts?

**Comment by floam** [comment]
`test` actually also works for floats in master, FYI.

**Comment by floam** [comment]
https://github.com/codeplea/tinyexpr/pull/38 implements logical operators, comparisons in tinyexpr. I think we want ours to do that.

**Comment by faho** [comment]
>However, removing the comparisons from math will break some existing fish scripts

We're explicitly doing a major release, so some breakage is expected. And I haven't found a single use of these in the wild.

Do you have any examples of fish things that break because of this?

**Comment by mojoTX** [comment]
@faho:  I noticed the issue with a weather plugin for oh-my-fish. Here's my latest pull request to just use `bc` directly for a floating point comparison, since that's guaranteed to be portable, and because performance isn't an issue in this case. 

https://github.com/oh-my-fish/plugin-weather/pull/24

**Comment by zanchey** [comment]
Version comparisons really should use sort's `-V` option.

**Comment by faho** [comment]
>sort's -V option

Unfortunately that's [not a thing on macos](https://ss64.com/osx/sort.html).

**Comment by floam** [comment]
It's done `sort -V` for a few years now, not sure exactly when it was added.

**Comment by faho** [comment]
>It's done sort -V for a few years now

I'd love to confirm that via an official source, but apparently Apple has taken down the man pages, and I've never found any proper changelog from them.

Any way, in this case it's still an improvement to test component-by-component instead of treating it as a float.

I initially forgot about "1.10". `echo "1.10 < 1.5" | bc` returns 1 (i.e. true). Version numbers are not floats.

**Comment by floam** [comment]
Here's the source for sort as included with 10.13, the man page it comes with is also there. You can go up a couple directories to find previous versions. 

https://opensource.apple.com/source/text_cmds/text_cmds-99/sort/

(You can go to the main page, https://opensource.apple.com, click different macOS versions, and the listing you see from them will tell you that e.g. 10.12 comes with `text_cmds` version 97.)

**Comment by floam** [comment]
So it looks like they had the sort without `--version-sort` on 10.12.
